invenio_records_resources/services/files/transfer.py

```diff
@@ -0,0 +1,156 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2022 CERN.
+#
+# Invenio-Records-Resources is free software; you can redistribute it and/or
+# modify it under the terms of the MIT License; see LICENSE file for more
+# details.
+
+"""Files transfer."""
+
+from abc import ABC, abstractmethod
+from enum import Enum
+
+from invenio_db import db
+from invenio_files_rest.errors import FileSizeError
+
+
+class TransferType(str, Enum):
+    """File type, it inherits from str to be JSON serializable.
+
+    LOCAL represents a file that is stored locally in the instance's storage.
+    EXTERNAL represents a file that needs to be fetched from an external storage
+             and saved locally.
+    LINKED represents a file that is stored externally and is linked to the record.
+    """
+
+    LOCAL = "S"  # S = Stored
+    EXTERNAL = "E"
+    LINKED = "L"
+
+    def __eq__(self, other):
+        """Equality test."""
+        return self.value == other
+
+    def __str__(self):
+        """Return its value."""
+        return self.value
+
+
+class BaseTransfer(ABC):
+    """Local transfer."""
+
+    def __init__(self, type):
+        """Constructor."""
+        self.type = type
+
+    @abstractmethod
+    def init_file(self, record, file_metadata):
+        """Initialize a file."""
+        pass
+
+    def set_file_content(self, record, file, file_key, stream, content_length):
+        """Set file content."""
+        bucket = record.bucket
+        size_limit = bucket.size_limit
+        if content_length and size_limit and content_length > size_limit:
+            desc = (
+                "File size limit exceeded."
+                if isinstance(size_limit, int)
+                else size_limit.reason
+            )
+            raise FileSizeError(description=desc)
+
+        record.files.create_obj(
+            file_key, stream, size=content_length, size_limit=size_limit
+        )
+
+    @abstractmethod
+    def commit_file(self, record, file_metadata):
+        """Commit a file."""
+        pass
+
+    # @abstractmethod
+    # def read_file_content(self, record, file_metadata):
+    #     """Read a file content."""
+    #     pass
+
+
+class LocalTransfer(BaseTransfer):
+    """Local transfer."""
+
+    def __init__(self):
+        """Constructor."""
+        super().__init__(TransferType.LOCAL)
+
+    def init_file(self, record, file_metadata):
+        """Initialize a file."""
+        uri = file_metadata.pop("uri", None)
+        if uri:
+            raise Exception("Cannot set URI for local files.")
```

Comment:
minor: probably we'll need a base `TransferException`, but maybe as a follow-up task in error handling.